maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java

```diff
@@ -49,45 +56,128 @@ class ReverseTreeRepositoryListener extends AbstractRepositoryListener {
         }
 
-        Artifact resolvedArtifact = event.getArtifact();
-        Artifact nodeArtifact = collectStepTrace.getNode().getArtifact();
-
-        if (isInScope(resolvedArtifact, nodeArtifact)) {
-            Dependency node = collectStepTrace.getNode();
-            ArrayList<String> trackingData = new ArrayList<>();
-            trackingData.add(node + " (" + collectStepTrace.getContext() + ")");
-            String indent = "";
-            ListIterator<DependencyNode> iter = collectStepTrace
-                    .getPath()
-                    .listIterator(collectStepTrace.getPath().size());
-            while (iter.hasPrevious()) {
-                DependencyNode curr = iter.previous();
+        Path trackingDir;
+        boolean missing = event.getFile() == null;
+        if (missing) {
+            // missing artifact - let's track the path anyway
+            File dir = event.getSession().getLocalRepository().getBasedir();
+            dir = new File(
+                    dir, event.getSession().getLocalRepositoryManager().getPathForLocalArtifact(event.getArtifact()));
+            trackingDir = dir.getParentFile().toPath().resolve(".tracking");
+        } else {
+            trackingDir = event.getFile().getParentFile().toPath().resolve(".tracking");
+        }
+
+        String baseName;
+        String ext = missing ? ".miss" : ".dep";
+        Path trackingFile = null;
+
+        String indent = "";
+        ArrayList<String> trackingData = new ArrayList<>();
+
+        if (collectStepTrace == null && plugin != null) {
+            ext = ".plugin";
+            baseName = plugin.getGroupId() + "_" + plugin.getArtifactId() + "_" + plugin.getVersion();
+            trackingFile = trackingDir.resolve(baseName + ext);
+            if (Files.exists(trackingFile)) {
+                return;
+            }
+
+            if (event.getArtifact() != null) {
+                trackingData.add(indent + event.getArtifact());
+                indent += "  ";
+            }
+            trackingData.add(indent + plugin.getGroupId() + ":" + plugin.getArtifactId() + ":" + plugin.getVersion());
+            indent += "  ";
+
+            InputLocation location = plugin.getLocation("");
+            if (location != null && location.getSource() != null) {
+                trackingData.add(indent + location.getSource().getModelId() + " (implicit)");
                 indent += "  ";
-                trackingData.add(indent + curr + " (" + collectStepTrace.getContext() + ")");
             }
-            try {
-                Path trackingDir =
-                        resolvedArtifact.getFile().getParentFile().toPath().resolve(".tracking");
-                Files.createDirectories(trackingDir);
-                Path trackingFile = trackingDir.resolve(collectStepTrace
+        } else if (collectStepTrace != null) {
+            baseName =
+                    collectStepTrace.getPath().get(0).getArtifact().toString().replace(":", "_");
```

Comment:
There is a generic to ID function, I would rather prefer that.

Reply:
Thanks for comment - I'll review on Monday.

Reply:
@michael-o after quick look ... can you point me to the right "to ID function"?

Reply:
Please see org.eclipse.aether.internal.impl.synccontext.named.GAVNameMapper.fileGav()

Reply:
Hmm, `org.eclipse.aether.artifact.AbstractArtifact#toString()` also uses extension and classifier, but `org.eclipse.aether.internal.impl.synccontext.named.GAVNameMapper#getArtifactName()` does not (and it's private, but using `nameLocks()` sounds like mixing concepts here - I don't deal with locks...)

Is there any other sanitizing method?

Reply:
https://github.com/apache/maven-resolver/blob/master/maven-resolver-util/src/main/java/org/eclipse/aether/util/artifact/ArtifactIdUtils.java#L28 maybe?

Reply:
Thanks, checking.
However current it's just toString + replace: https://github.com/apache/maven/blob/maven-3.9.1/maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java#L81-L86

Reply:
Yes, that is not (yet) flexible, use a `replace()` for now.